File: 200-number-of-islands/number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int count  = 0;
        int rowLen = grid.length;
        int colLen = grid[0].length;

        for(int i=0;i<rowLen;i++) {
            for(int j=0;j<colLen;j++) {
                if(grid[i][j]=='1') {
                    count++;

                    Queue<int[]> queue = new LinkedList<>();

                    queue.add(new int[]{i,j});

                    while(!queue.isEmpty()) {
                        int[] elem = queue.poll();

                        //grid[elem[0]][elem[1]] = '0';

                        if(elem[0]-1>=0 && grid[elem[0]-1][elem[1]]=='1') {
                            grid[elem[0]-1][elem[1]]='0';
                            queue.add(new int[]{elem[0]-1,elem[1]});
                        }
                        if(elem[0]+1 < rowLen && grid[elem[0]+1][elem[1]]=='1') {
                            grid[elem[0]+1][elem[1]]='0';
                            queue.add(new int[]{elem[0]+1,elem[1]});
                        }
                        if(elem[1]-1>=0 && grid[elem[0]][elem[1]-1]=='1') {
                            grid[elem[0]][elem[1]-1]='0';
                            queue.add(new int[]{elem[0],elem[1]-1});
                        }
                        if(elem[1]+1 < colLen && grid[elem[0]][elem[1]+1]=='1') {
                            grid[elem[0]][elem[1]+1]='0';
                            queue.add(new int[]{elem[0],elem[1]+1});
                        }
                        
                    }

                    
                }
            }
            
        }

        return count;
        
    }

}

File: 200-number-of-islands/number-of-islands.cpp
class Solution {
public:
    void dfs(vector<vector<char>>& grid, int r, int c) {
        int row_size = grid.size();
        int col_size = grid[0].size();

        if(r<0 || r>row_size-1 || c<0 || c>col_size-1) {
            return;
        }

        if(grid[r][c]=='0') {
            return;
        }

        grid[r][c]='0';

        dfs(grid, r-1,c);
        dfs(grid, r+1, c);
        dfs(grid, r, c-1);
        dfs(grid, r, c+1);
    }
    int numIslands(vector<vector<char>>& grid) {
        int count=0;

        int row_size = grid.size();
        int col_size = grid[0].size();

        for(int i=0;i<row_size;i++) {
            for(int j=0;j<col_size;j++) {
                if(grid[i][j]=='1') {
                    count++;
                    dfs(grid, i, j);
                }
            }
        }
        return count;
    }

};